Handle 1-4TiB provider config and upscaling

raaizik opened the pull request:
Resolves [Faas-44](https://jsw.ibm.com/browse/FAAS-44), [Faas-45](https://jsw.ibm.com/browse/FAAS-45):

- Adds support for `UsableCapacityInTiB` config knob with values 1-4TiB (units included)
- Performs up-scaling and prevents down-scaling

TODOs:
- [ ] Check regarding optimal perf of different setups of {capacity, OSD size} combos 
- [x] Check use case of non-integer `UsableCapacityInTiB` is handled correctly
- [x] Test manually

![image](https://github.com/red-hat-storage/managed-fusion-agent/assets/132667934/d8fa31c3-fc0d-4d70-b862-26e9c35a3ee7)

raaizik committed:
Fix for calculating count and setting OSD size

Downscaling is prevented via both OSDSizeInTiB and Count knobs

Initial commit

OSDSizeInTiB is being picked up from DataPVCTemplate's requests of the StorageCluster

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Revert "Fix for calculating count and setting OSD size"

This reverts commit 58f8451989a1ad4e6257bec401960758513671e3.

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

1st draft

- Handles 1-4 TiB config
- Prevents downscaling

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Update datafoundation.go

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Update datafoundation.go

Fix to updatedSCSize

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Add const VS upper bound

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Fix upper case

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Rename VS upper bound

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Fix and update template

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Works

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

rchikatw commented on controllers/datafoundation.go:
```diff
@@ -271,25 +272,48 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
-		desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB) / templates.OSDSizeInTiB))
+		// get the capacity in TiB planned to be consumed by user
+		storageClusterSize := math.Ceil(float64(r.spec.UsableCapacityInTiB))
+		if storageClusterSize <= templates.VerticalScalerUpperBoundInTiB {
```
No need to use an extra variable because it's been used only in one place.

rchikatw reviewed (commented):
Also squash your commits in a single commit.

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -271,25 +272,48 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
-		}
+			// Count remains 1
+			// todo need to confirm perf of all combos of Count and OSD Sizes (i.e. 2 --> count = 1 and OSD Size = 2 OR count = 2 and OSD Size = 1)
+			r.storageCluster.Spec = desiredStorageCluster.Spec
```
We should not update the current storage cluster spec, we should update the template instance which override the spec (later in the function)

nb-ohad commented on controllers/datafoundation.go:
A general comment about this change. 

The way I see it we need to calculate only 2 values:
```
desiredDeviceSetCount := int(
    math.Ceil(
        math.Max(float64(r.spec.UsableCapacityInTiB), VerticalScalerUpperBoundInTiB) / templates.OSDSizeInTiB
    )
)
```

and 

```
OSDSize  := math.Min(float64(r.spec.UsableCapacityInTiB), VerticalScalerUpperBoundInTiB)
```
And we only need one protection guard, to make sure the new desired size is bigger than the actual size.


This means that code changes here should be minor at best and the flow should stay similar to the original one

raaizik replied:
Don't you mean the denominator is a function of `UsableCapacityInTiB`, and not the constant `templates.OSDSizeInTiB`?, Meaning:

```go
desiredDeviceSetCount := int(
    math.Ceil(
        math.Max(float64(r.spec.UsableCapacityInTiB), VerticalScalerUpperBoundInTiB) / OSDSizeInTiB
    )
)
``` 
where 
```go
OSDSizeInTiB := math.Min(float64(r.spec.UsableCapacityInTiB), templates.VerticalScalerUpperBoundInTiB)

```

nb-ohad replied:
What I am trying to say is that given that the max OSD size is 4 then for every given valid cluster size we can compute the OSD size and the OSD count (desiredDeviceSetCount):
```
OSD size = min(cluster size, 4)
OSD count = ceil(max(cluster size, 4) / 4)
```

Coming to think about it, there is no need to have `templates.VerticalScalerUpperBoundInTiB` at all as its value is always going to match `templates.OSDSizeInTiB`

nb-ohad replied:
Looking at my last comment, I think we can even simplify the OSD count formula
``` 
ceil(max(cluster size, 4) / 4) => ceil(max(cluster size / 4, 1)) => ceil(clusterSize / 4) 
```
This means the actual formals are:
```
OSD size = min(cluster size, 4)
OSD count = ceil(cluster size / 4)
```

raaizik replied:
I think my suggestion is readable. IMO there also needs to be a differentiation between the initialization value and the one associated with the vertical scaler, therefore:
```Go
OSDSizeInTiB                  = 1
VerticalScalerUpperBoundInTiB = 4
```

nb-ohad replied:
Resolving in favor of a newer review comment.

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -271,12 +273,21 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
-		// Convert the desired size to the device set count based on the underlaying OSD size
-		desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB) / templates.OSDSizeInTiB))
+		// get the desired capacity in TiB planned to be consumed by user in range 1-4TiB
+		desiredOSDSizeInTiB := math.Min(math.Ceil(float64(r.spec.UsableCapacityInTiB)), templates.VerticalScalerUpperBoundInTiB)
```
Something does not make sense with 
```
desiredOSDSizeInTiB := math.Min(math.Ceil(float64(r.spec.UsableCapacityInTiB)), templates.VerticalScalerUpperBoundInTiB)
```

Why are we basing desired state on actual state (`r.spec.*`)?


Also `ceil(float(int))` will just return the `int` as a float. 
This means that the above statement can be simplified to
```suggestion
desiredOSDSizeInTiB := math.Min(
    float64(r.spec.UsableCapacityInTiB), 
    float64(templates.VerticalScalerUpperBoundInTiB),
)
```

raaizik replied:
> Why are we basing desired state on actual state (r.spec.*)?

That was my exact question, but I just followed the logic (that doesn't make sense) that existed prior to my change [controllers/datafoundation.go#L275](https://github.com/red-hat-storage/managed-fusion-agent/blob/main/controllers/datafoundation.go#L275)
```Go
desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB) / templates.OSDSizeInTiB))
```

nb-ohad replied:
The `Ceil` on the orignal call was around the division where it made sense that we will round up the result. 
In the new code it is just a `Ceil` on a float conversion. That is the difference

raaizik replied:
Note I was referring to the questionable use of `r.spec` in the old code and nothing else. In the new code it does make sense to base the desired state on the above since this is basically the flow required to update this variable, polled via the runtime object

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -271,12 +273,21 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
 
-		// Get the storage device set count of the current storage cluster
+		// Convert the desired size to the device set count based on the underlying OSD size
+		desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB)) / desiredOSDSizeInTiB)
```
Very similar questions here.
Why are we basing desired state on actual state (r.spec.*)?

Also, I believe there is a bug here. Converting to `int` will lose the friction not round it up. The actual statement should be
```suggestion
		desiredDeviceSetCount := math.Ceil(float64(r.spec.UsableCapacityInTiB) / desiredOSDSizeInTiB)
```

raaizik replied:
I guess you meant the casting should wrap the entire expression (it is after all an integer):
```Go
desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB) / desiredOSDSizeInTiB))
```

nb-ohad replied:
Yep, that makes sense

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -271,12 +273,21 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
 		currDeviceSetCount := 0
+		currOSDSizeInTiB := 0.0
 		if desiredStorageDeviceSet := findStorageDeviceSet(r.storageCluster.Spec.StorageDeviceSets, defaultDeviceSetName); desiredStorageDeviceSet != nil {
+			// Avoid overflow
+			if len(strconv.Itoa(int(desiredStorageDeviceSet.DataPVCTemplate.Spec.Resources.Requests.Storage().Value()))) <= int(resource.Tera) {
+				return fmt.Errorf("calculating the current OSD size in TiB might result in overflow")
+			}
+			// Get rounded up value and convert to actual
+			currOSDSizeInTiB = float64(desiredStorageDeviceSet.DataPVCTemplate.Spec.Resources.Requests.Storage().ScaledValue(resource.Tera) - 1)
```
There is no need to convert this into an int or a float. I believe that resource quantity provides methods so they can be compared.

```suggestion
currOSDSizeInTiB := desiredStorageDeviceSet.DataPVCTemplate.Spec.Resources.Requests.Storage()
```

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -287,7 +298,8 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
 
-		// Prevent downscaling by comparing count from secret and count from storage cluster
+		// Prevent downscaling by comparing OSD size and count from secret and from storage cluster
+		r.setDeviceSetOSDSize(ds, desiredOSDSizeInTiB, currOSDSizeInTiB)
```
I don't see a point in having this separate from `setDevicesetCount`. Mainly as we need to protect against downscale as a result of these pair of parameters configured together. 

You might want to rename `setDeviceSetCount` to `setDeviceSetProperties` and have the entire logic in one place

raaizik replied:
I was about to

nb-ohad commented on datafoundation/templates/storagecluster.go:
```diff
@@ -27,8 +27,9 @@ import (
 const (
-	OSDSizeInTiB            = 4
-	backingStorageClassName = "default-ocs-storage-class"
+	OSDSizeInTiB                  = 1
+	VerticalScalerUpperBoundInTiB = 4
```
I am fine with you wanting to change the name of the upper limit of allowed OSD size but

1. OSDSizeInTiB has no use anymore so we should not keep it
2. the new name `VerticalScalerUpperBoundInTiB` need to identify itself as belonging to the OSD value because it is an export of the entire `templates` package. 

maybe:
```suggestion
	OSDSizeUpperBoundInTiB = 4
```

raaizik replied:
> OSDSizeInTiB has no use anymore so we should not keep it

[datafoundation/templates/storagecluster.go#L143](https://github.com/red-hat-storage/managed-fusion-agent/blob/18f7d0634c6e8aa9b27e25c05ed85da1f1219bb8/datafoundation/templates/storagecluster.go#L143)
```Go
"storage": resource.MustParse(fmt.Sprintf("%dTi", OSDSizeInTiB)),
```
As I mentioned this is basically the initialization part. It can't be initialized to 4 as it was since then we won't be able to have the full range of 1-4TiB due to the downscaling check. That was the point I was making about being able to differentiate between initialization and scaling.

Edit: omitted `OSDSizeInTiB`

raaizik committed:
Converted to comply with Quantity

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Converted to Quantity and refactored

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Renaming and casting correction

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Update controllers/datafoundation.go

Co-Authored-By: nb-ohad <ohad.mitrani@noobaa.com>

rewantsoni commented on controllers/datafoundation.go:
```diff
@@ -271,12 +272,27 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
 
-		// Get the storage device set count of the current storage cluster
+		// get the desired capacity in TiB planned to be consumed by user in range 1-4TiB
+		desiredOSDSizeInTiBQuantity := OSDSizeUpperBoundInTiBQuantity
+		// Set desired to be the minimum
+		if UsableCapacityInTiBQuantity.Cmp(OSDSizeUpperBoundInTiBQuantity) <= 0 {
+			desiredOSDSizeInTiBQuantity = UsableCapacityInTiBQuantity
+		} else {
+			desiredOSDSizeInTiBQuantity = OSDSizeUpperBoundInTiBQuantity
+		}
```
We don't need a else here

rewantsoni commented:
- Can we merge all the commits to be a single commit?
- Also, could you remove WIP from the PR title?

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -50,7 +51,7 @@ const (
 
 type dataFoundationSpec struct {
-	UsableCapacityInTiB     int
+	UsableCapacityInTiB     resource.Quantity
```
This is an external contract/api/input in the form of a CR interface. We cannot change the type here as it is expected to just be a number. lets keep this as an int (we don't have a choice)

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -262,7 +263,10 @@ func (r *dataFoundationReconciler) reconcileOnboardingValidationSecret() error {
 	r.Log.Info("Reconciling StorageCluster")
 
-	if r.spec.UsableCapacityInTiB <= 0 {
+	UsableCapacityInTiB := r.spec.UsableCapacityInTiB
```
As of the above comment. we will need to translate from int to quantity only once here

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -262,7 +263,10 @@ func (r *dataFoundationReconciler) reconcileOnboardingValidationSecret() error {
+
+	if UsableCapacityInTiB.CmpInt64(0) <= 0 {
 		return fmt.Errorf("invalid or missing field: UsableCapacityInTiB, provided value is %v", r.spec.UsableCapacityInTiB)
 	}
```
The original intention for this was a sanity check on the provided input (the datafoundation spec) and as such was the first step of the reconcile function, before anything else. 

Is there a reason to change it to work against a different variable and against a quantity and not an int (the spec is given as an int)

nb-ohad commented on datafoundation/templates/storagecluster.go:
```diff
@@ -27,10 +25,12 @@ import (
 const (
-	OSDSizeInTiB            = 4
-	backingStorageClassName = "default-ocs-storage-class"
+	OSDSizeUpperBoundInTiBUnscaled = 4398046511104
+	backingStorageClassName        = "default-ocs-storage-class"
 )
 
+var OSDSizeUpperBoundInTib = resource.NewQuantity(OSDSizeUpperBoundInTiBUnscaled, resource.BinarySI)
+
```
If we cannot do an actual cost, can we mitigate the chance that someone will change the var value from the outside by using a functions?

```suggestion
	backingStorageClassName        = "default-ocs-storage-class"
)

func GetOSDSizeUpperBoundInTib() resource.Quantity {
    return resource.NewQuantity(4398046511104, resource.BinarySI)
}
```

raaizik replied:
Switched back to `MustParse` in function scope

raaizik committed:
Parse ints

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

nb-ohad commented on controllers/datafoundation.go:
```diff
@@ -271,24 +272,34 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
 		}
 
+		// Convert the desired size to the device set count based on the underlying OSD size
+		desiredDeviceSetCount := int(math.Ceil(UsableCapacityInTiB.AsApproximateFloat64() / desiredOSDSizeInTiB.AsApproximateFloat64()))
```
No need to transform back to bytes and lose precision, you can use `ScaledValue` as a better alternative.
```suggestion
		desiredDeviceSetCount := int(math.Ceil(
			float(UsableCapacityInTiB.ScaledValue("Tera")) / float(desiredOSDSizeInTiB.ScaledValue("Tera"))
		))
```

nb-ohad replied:
@raaizik This is not a must, if you feel the code is good as is we can merge it as is

raaizik replied:
@nb-ohad At some point in the future we would probably need to support non-tera values, no?

nb-ohad replied:
Storage provisioning for an entire cluster (not a volume provided by the cluster) is provisioned at the minimum GiBs levels, but in most cases at the TiB level. I would even make a case that our code should prevent allowing clusters to be provisioned at a unit below the GiB unit. 

If you feel that you prefer to translate everything to GiBs, I am ok with that. The main thing is that I would like to eliminate the imprecisions that come with translating to very big floats (in this case, translating to bytes)

raaizik committed:
Update datafoundation.go

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Update datafoundation.go

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Avoid loss of percision

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>

Update datafoundation.go

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>
Co-Authored-By: nb-ohad <ohad.mitrani@noobaa.com>
